---
title: "Erika_Final_Project"
author: "Erika Mehrhoff"
date: "4/24/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(echo = TRUE)
#library(Gviz)
#library(ggpubr)
library(ggplot2)
library(tidyverse)
library(GenomicRanges)
#library(ComplexHeatmap)
library(circlize)
library(colorRamps)
source("../../../../util/my_class_functions.R")
source("../../../../util/_setup.R")
```

# Your mission should you choose to accept it 
There is an enormous amount of data in ENCODE that 
needs to be analyzed. This includes 480 unique DBP
ChIPseq datasets. To facilitate your mission we provide you
1,066 peak files that have been run through the NF_CORE chipseq
pipeline. 

You will find all the .broadPeak peak files here:
/scratch/Shares/rinnclass/CLASS_2023/data/data/peaks

Use what you have learned in class to uncover
new results from this daunting data set. 
# Good luck the genome depends on you!


# first you might want to think about your organization
# 1) create a directory in your class folder for these analyses
    # -> folder name
          -> util/
            -> scripts.R
          -> analysis/
            -> 01_consensus_peaks
              -> .RMD
              -> results/
              -> figures
            -> 02_clustering etc....
  
  
# Consensus peaks
# Num_peaks_df
# Peak_occurence_df
# lncRNA and or mRNA promoters / gene annotations
# SAVE THIS !! save(object_name, nextobject, ...., file = "where you save this.Rdata")
# load(file.Rdata)


# First I want to analyze the large data for general properties 
# Import broad peak files and find consensus peaks between replicates
```{r}
# filepath to import peaks
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/erme3555"
peak_path <- "CLASS_2023/CLASSES/05_R_analyses/analysis/00_consensus_peaks"
broadpeakfilepath <- "/scratch/Shares/rinnclass/CLASS_2023/data/data/peaks"

# using import peaks to import .broadPeak files (~10min)
peak_list <- import_peaks(consensus_file_path = broadpeakfilepath)

# Creating unique DBP object for create_consensus_peaks_from_reduced
dbp <- unique(sapply(names(peak_list), function(x) {
   unlist(strsplit(x, "_"))[1]
})) # remove replicate number

# now run our function consensus_from_reduced
consensus_list <- lapply(dbp, consensus_from_reduced, peak_list)

# adding names to the GRange list
names(consensus_list) <- dbp

save(consensus_list, file = "results/consensus_list.RData")

load("results/consensus_list.RData", verbose = T)

# creating list of num_peaks per dbp
num_peaks <- sapply(consensus_list, length) # vector of number of peaks for each protein

# Filter out any chip data less 1,000 peaks == filtered consensus peaks
filtered_consensus_list <- consensus_list[sapply(consensus_list, length) > 1000]

# saving 
save(filtered_consensus_list, file = "results/filtered_consensus_list.RData")

# keeping track of DBPs lost
lost_dbps <- names(consensus_list[sapply(consensus_list, length) < 1000]) %>% as.data.frame()

# saving 
write.table(lost_dbps, "results/lost_dbps.csv")

# exporting filtered_consensus_peaks
for(i in 1:length(filtered_consensus_list)) {
  rtracklayer::export(filtered_consensus_list[[i]], 
                      paste0("results/filtered_consensus_peaks/", 
                             names(filtered_consensus_list)[i], 
                             "_filtered_consensus_peaks.bed"))
}
```
# Results:
We started with chip data on 486 DNA binding proteins, however after filtering out any chip data less 1,000 peaks there were 430 DNA binding proteins remaining. 

# Prep Genome features
```{r}
# loading in genome features
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/gencode.v32.annotation.gtf")

# gencode genes
gencode_genes <- gencode_gr[gencode_gr$type == "gene"] 

# mrna_genes
mrna_genes <- gencode_genes[gencode_genes$gene_type %in% "protein_coding"]

# lncrna_genes
lncrna_genes <- gencode_genes[gencode_genes$gene_type %in% "lncRNA"] 

# mrna_lncrna_genes
mrna_lncrna_genes <- gencode_genes[gencode_genes$gene_type %in% c("protein_coding","lncRNA")]

# lncrna_mrna_promoters
lncrna_mrna_promoters <- promoters(mrna_lncrna_genes, upstream = 1000, downstream = 1000)

# lncrna_gene_ids
lncrna_gene_ids <- mrna_lncrna_genes$gene_id[mrna_lncrna_genes$gene_type == "lncRNA"]

# mrna_gene_ids
mrna_gene_ids <-mrna_lncrna_genes$gene_id[mrna_lncrna_genes$gene_type == "protein_coding"]
```

# Comparing number of peaks to genome coverage
```{r}
num_peaks_df <- data.frame("dbp" = names(filtered_consensus_list),
                           "num_peaks" = sapply(filtered_consensus_list, length))

# total genome covered by peaks
num_peaks_df$total_peak_length <- sapply(filtered_consensus_list, function(x) sum(width(x)))

# How does peak number and genome coverage compare
ggplot(num_peaks_df,
       aes(x = num_peaks, y = total_peak_length)) +
  xlab("Peaks per DBP") +
  ylab("Total Genome Coverage") +
  ggtitle("Relationship Between Number of DBP Peaks and Genome Coverage")+
  geom_point() +
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=FALSE, formula = 'y ~ x',
              color = "#a8404c")
```
#Results: There is a positive linear relationship between number of peaks and the total genome coverage.

# Distribution of promoter overlaps vs gene-bodies
```{r}
# What is the distribution of promoter overlaps versus gene-bodies (hint hist)

# creating number of promoter overlaps entry
promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_list, type = "counts")

# summing rows to get total number of promoter overlaps
num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)

# Finding overlaps with gene_bodies (will take a few minutes again)
genebody_peak_counts <- count_peaks_per_feature(mrna_lncrna_genes, 
                                                filtered_consensus_list, 
                                                type = "counts")

# All gene bodies overlaps
num_peaks_df$peaks_overlapping_genebody <- rowSums(genebody_peak_counts)

hist(num_peaks_df$peaks_overlapping_promoters)
hist(num_peaks_df$peaks_overlapping_genebody)
```
# Results: The distribution of peaks overlapping promoters is different than peaks overlapping gene-bodies. The distribution of peaks overlapping gene-bodies is right skewed. 

# Super Binders
```{r}
# Make a list of genes that are "super binders" 
#running count_peaks_per_feature
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_list, 
                                               type = "occurrence")

# Let's double check that all lncrna & mrna genes are accounted for:
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))

# Now let's use the 'data.frame()' fucntion. Set up a bunch of colnames and populate them.
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "gene_name" = lncrna_mrna_promoters$gene_name,
                                "gene_type" = lncrna_mrna_promoters$gene_type,
                                "chr" = lncrna_mrna_promoters@seqnames,   
                                "1kb_up_tss_start" = lncrna_mrna_promoters@ranges@start,
                                "strand" = lncrna_mrna_promoters@strand,
                                "number_of_dbp" = colSums(promoter_peak_occurence))

# Look at the distribution of binding and pick cut off for super binders
ggplot(peak_occurence_df, aes(x = number_of_dbp)) +
geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes") 

# list of super binders
super_binders <- filter(peak_occurence_df, number_of_dbp >= 200)
super_binders$gene_name
# There are 11722 super binders (out of 36814)

# Is there a type of gene ontology associated with them versus the others?
# column for type of gene or google genes?
  
  
# Is there a difference in mRNA and lncRNA promoter overlaps?
reg_binders <- filter(peak_occurence_df, number_of_dbp < 200)
rb_lnc <- nrow(reg_binders[reg_binders$gene_type=="lncRNA",])
rb_mr <- nrow(reg_binders[reg_binders$gene_type=="protein_coding",])
sb_lnc <- nrow(super_binders[super_binders$gene_type=="lncRNA",])
sb_mr <- nrow(super_binders[super_binders$gene_type=="protein_coding",])

rnames <- c("mRNA","lncRNA")
cnames <- c("Regular","Super Binder")
matrix(data= c(rb_mr, rb_lnc,sb_mr , sb_lnc ), ncol=2, dimnames=list(rnames,cnames))

# Do lncRNAs also have super-binding promoters?
```
# Results: Super-binders are more associated with mRNA than lncRNA, which is the opposite in the reguilar binders. There were 9,201 mRNA superbinding promoters and 2,521 lncRNA superbinding promoters.  

Transcription Factors
```{r}
# How many of these proteins are TFs? What is the most represented type of DBD?

# reading in TF annotations 
human_tfs <- readxl::read_excel("../../analysis/01_create_consensus_peaks/results/TF_annotations.xlsx",
                                sheet = 2, skip = 1)

# let's rename the 4th column to indicate if it is a TF.
names(human_tfs)[4] <- "is_tf"

# now let's intersect gene names that are in our ChIP data and has TF identity.
length(which(tolower(num_peaks_df$dbp) %in% tolower(human_tfs$Name)))
# 407 of the 430 have matching gene_names - not bad

human_tfs <- human_tfs[tolower(human_tfs$Name) %in% tolower(num_peaks_df$dbp), 1:4]
# adding new column names
names(human_tfs) <- c("ensembl_id",
                      "dbp",
                      "dbd",
                      "tf")

# merging into num_peaks_df
num_peaks_df <- merge(num_peaks_df, human_tfs, all.x = T)

# count up which dbp are TFs
nrow(num_peaks_df[num_peaks_df$tf=="Yes",])
nrow(num_peaks_df[num_peaks_df$tf=="No",])

# most represented type of DBD
num_peaks_df$dbd <- as.factor(num_peaks_df$dbd)
names(which.max(table(num_peaks_df$dbd)))
```
# How many of these proteins are TFs? What is the most represented type of DBD?
# Results: 367 of the proteins are transcription factors (85%), and the other 86 are not. The most represented type of DBD is C2H2 ZF.

# Let's see how similar our genes are to each other
# Clustering 
```{r}

# What genes make sense to cluster together create an image (hint Pol II makes RNA)

# Find a cluster of genes your interested in and can figure out what those genes do -- are there unknown genes i there too? If so maybe one could hypothesize they have a similar function to the known genes (Hint ZNFs)

# if we cluster by lncRNA and mRNA separately what are some similarities and differences?

```
# Result: ZNFX seems to be similar to geneY and I hypothesize it might involved in A

# Metaplots
```{r}

# Let's look at the metaplot for all DBPs on lncRNA and mRNA promoters seperately (hint facet wrap).

# Which genes seem to have a difference in where they bind on promoters between lncRNA and mRNA promoters


# Make a metaplot of DBPS only on Super-binders versus regular promoters ...
```


# RNAseq expression

```{r}

# What is the relationship between number of DBPS bound on a promoter versus RNA output (hint TPM)

# Let's make a heatmap of genes that are variable across samples 

# Which subfraction of the cell has the highest expression.

# Let's make a heatmap of nuclear versus cytoplasmic expression

# How many lncRNA and mRNA genes are sig in nuclear or cyto

# What is/are the most nuclear mRNA(s) -- is there a type of gene ontology associated with them?

# If we zoom in on high binding promoters (> 200 DBPs) are there any that don't have any expression?

```
# Result in file.csv I have all the super binders that don't express RNA